File: ios/Classes/FlutterMediaLoadRequestDataHelper.swift
//
//  FlutterMediaLoadRequestDataHelper.swift
//  flutter_cast_framework
//
//  Created by Gianluca Paradiso on 19/11/21.
//

import Foundation
import GoogleCast

func getFlutterMediaStatus(mediaStatus: GCKMediaStatus?) -> MediaStatus {
    let result = MediaStatus()
    
    if (mediaStatus == nil) {
        return result
    }
    
    result.isPlayingAd = NSNumber(value: mediaStatus?.playingAd ?? false)
    result.mediaInfo = getFlutterMediaInfo(mediaInfo: mediaStatus?.mediaInformation)
    result.playerState = getFlutterPlayerState(playerState: mediaStatus?.playerState)
    result.adBreakStatus = getFlutterAdBreakStatus(adBreakStatus: mediaStatus?.adBreakStatus)
    
    return result
}

func getFlutterAdBreakStatus(adBreakStatus: GCKAdBreakStatus?) -> AdBreakStatus {
    let result = AdBreakStatus()
    
    if (adBreakStatus == nil) {
        return result
    }
    
    result.adBreakId = adBreakStatus?.adBreakID
    result.adBreakClipId = adBreakStatus?.adBreakClipID
    result.whenSkippableMs = getFlutterWhenSkippableMs(whenSkippable: adBreakStatus?.whenSkippable)
    
    return result
}

func getFlutterWhenSkippableMs(whenSkippable: TimeInterval?) -> NSNumber {
    if (whenSkippable == nil || whenSkippable?.isNaN == true || whenSkippable?.isInfinite == true) {
        return 0
    }
    
    let whenSkippableSecs = whenSkippable ?? 0
    let whenSkippableMs = Int(whenSkippableSecs * 1000)
    
    return NSNumber(value: whenSkippableMs)
}

func getFlutterPlayerState(playerState: GCKMediaPlayerState?) -> PlayerState {
    switch playerState {
    case .unknown:
        return PlayerState.unknown
    case .idle:
        return PlayerState.idle
    case .playing:
        return PlayerState.playing
    case .paused:
        return PlayerState.paused
    case .buffering:
        return PlayerState.buffering
    case .loading:
        return PlayerState.loading
    default:
        return PlayerState.unknown
    }
}

func getFlutterMediaInfo(mediaInfo: GCKMediaInformation?) -> MediaInfo {
    let result = MediaInfo()
    
    if (mediaInfo == nil) {
        return result
    }
    
    result.contentId = mediaInfo?.contentID
    result.contentType = mediaInfo?.contentType
    result.customDataAsJson = "\(mediaInfo?.customData ?? {})"
    result.mediaMetadata = getFlutterMediaMetadata(mediaMetadata: mediaInfo?.metadata)
    result.mediaTracks = getFlutterMediaTracks(mediaTracks: mediaInfo?.mediaTracks)
    result.streamDuration = getStreamDuration(streamDuration: mediaInfo?.streamDuration)
    result.streamType = getFlutterStreamType(streamType: mediaInfo?.streamType)
    result.adBreakClips = getFlutterAdBreakClips(adBreakClips: mediaInfo?.adBreakClips)
    
    return result
}

func getStreamDuration(streamDuration: TimeInterval?) -> NSNumber {
    if (streamDuration == nil || streamDuration?.isNaN == true || streamDuration?.isInfinite == true) {
        return 0
    }
    
    let durationRounded = streamDuration?.rounded() ?? 0
    let durationInt = Int(durationRounded)
    return NSNumber(value: durationInt)
}

func getFlutterStreamType(streamType: GCKMediaStreamType?) -> StreamType {
    switch streamType {
    case .buffered:
        return StreamType.buffered
    case .live:
        return StreamType.live
    case .unknown:
        return StreamType.invalid
    case nil:
        return StreamType.invalid
    case .some(.none):
        return StreamType.invalid
    @unknown default:
        return StreamType.invalid
    }
}

func getFlutterMediaTracks(mediaTracks: [GCKMediaTrack]?) -> [MediaTrack]? {
    if (mediaTracks == nil) {
        return nil
    }
    
    var result = [MediaTrack]()
    mediaTracks?.forEach({ (track: GCKMediaTrack) in
        let resultTrack = getFlutterMediaTrack(mediaTrack: track)
        result.append(resultTrack)
    })
    return result
}

func getFlutterMediaTrack(mediaTrack: GCKMediaTrack) -> MediaTrack {
    let result = MediaTrack()
    
    result.contentId = mediaTrack.contentIdentifier
    result.id = NSNumber(value: mediaTrack.identifier)
    result.language = mediaTrack.languageCode
    result.name = mediaTrack.name
    result.trackType = getFlutterTrackType(trackType: mediaTrack.type)
    result.trackSubtype = getFlutterTrackSubtype(trackSubtype: mediaTrack.textSubtype)
    
    return result
}

func getFlutterAdBreakClips(adBreakClips: [GCKAdBreakClipInfo]?) -> [AdBreakClipInfo]? {
    if (adBreakClips == nil) {
        return nil
    }
    
    var result = [AdBreakClipInfo]()
    adBreakClips?.forEach({ (info: GCKAdBreakClipInfo) in
        let resultInfo = getFlutterAdBreakClipInfo(adBreakClipInfo: info)
        result.append(resultInfo)
    })
    return result
}

func getFlutterAdBreakClipInfo(adBreakClipInfo: GCKAdBreakClipInfo) -> AdBreakClipInfo {
    let result = AdBreakClipInfo()
    
    result.id = adBreakClipInfo.adBreakClipID
    result.title = adBreakClipInfo.title
    result.contentId = adBreakClipInfo.contentID
    result.contentUrl = adBreakClipInfo.contentURL?.absoluteString
    result.clickThroughUrl = adBreakClipInfo.clickThroughURL?.absoluteString
    result.durationMs = getStreamDuration(streamDuration: adBreakClipInfo.duration)
    result.imageUrl = adBreakClipInfo.posterURL?.absoluteString
    result.mimeType = adBreakClipInfo.mimeType
    result.whenSkippableMs = getFlutterWhenSkippableMs(whenSkippable: adBreakClipInfo.whenSkippable)
    
    return result
}

func getFlutterTrackType(trackType: GCKMediaTrackType) -> TrackType {
    switch trackType {
    
    case .unknown:
        return TrackType.unknown
    case .text:
        return TrackType.text
    case .audio:
        return TrackType.audio
    case .video:
        return TrackType.video
    @unknown default:
        return TrackType.unknown
    }
}

func getFlutterTrackSubtype(trackSubtype: GCKMediaTextTrackSubtype) -> TrackSubtype {
    switch trackSubtype {
    
    case .unknown:
        return TrackSubtype.unknown
    case .subtitles:
        return TrackSubtype.subtitles
    case .captions:
        return TrackSubtype.captions
    case .descriptions:
        return TrackSubtype.descriptions
    case .chapters:
        return TrackSubtype.chapters
    case .metadata:
        return TrackSubtype.metadata
    @unknown default:
        return TrackSubtype.unknown
    }
}

func getFlutterMediaMetadata(mediaMetadata: GCKMediaMetadata?) -> MediaMetadata {
    let result = MediaMetadata()
    
    result.mediaType = getFlutterMediaType(mediaType: mediaMetadata?.metadataType)
    result.strings = [:]
    result.webImages = []
    
    mediaMetadata?.allKeys().forEach({ key in
        let flutterKey = getFlutterMediaMetadataKey(mediaMetadataKey: key)
        result.strings?[flutterKey] = mediaMetadata?.string(forKey: key)
    })
    
    mediaMetadata?.images().forEach({ (imageAny: Any) in
        let image = imageAny as? GCKImage
        if (image == nil) {
            return
        }
        
        let webImage = WebImage()
        webImage.url = image?.url.absoluteString
        result.webImages?.append(webImage)
    })
    
    return result
}

func getFlutterMediaMetadataKey(mediaMetadataKey: String) -> String {
    switch mediaMetadataKey {
    case kGCKMetadataKeyAlbumArtist:
        return getEnumLabelFor(key: MediaMetadataKey.albumArtist)
    case kGCKMetadataKeyAlbumTitle:
        return getEnumLabelFor(key: MediaMetadataKey.albumTitle)
    case kGCKMetadataKeyArtist:
        return getEnumLabelFor(key: MediaMetadataKey.artist)
    case kGCKMetadataKeyBookTitle:
        return getEnumLabelFor(key: MediaMetadataKey.bookTitle)
    case kGCKMetadataKeyBroadcastDate:
        return getEnumLabelFor(key: MediaMetadataKey.broadcastDate)
    case kGCKMetadataKeyChapterNumber:
        return getEnumLabelFor(key: MediaMetadataKey.chapterNumber)
    case kGCKMetadataKeyChapterTitle:
        return getEnumLabelFor(key: MediaMetadataKey.chapterTitle)
    case kGCKMetadataKeyComposer:
        return getEnumLabelFor(key: MediaMetadataKey.composer)
    case kGCKMetadataKeyCreationDate:
        return getEnumLabelFor(key: MediaMetadataKey.creationDate)
    case kGCKMetadataKeyDiscNumber:
        return getEnumLabelFor(key: MediaMetadataKey.discNumber)
    case kGCKMetadataKeyEpisodeNumber:
        return getEnumLabelFor(key: MediaMetadataKey.episodeNumber)
    case kGCKMetadataKeyHeight:
        return getEnumLabelFor(key: MediaMetadataKey.height)
    case kGCKMetadataKeyLocationLatitude:
        return getEnumLabelFor(key: MediaMetadataKey.locationLatitude)
    case kGCKMetadataKeyLocationLongitude:
        return getEnumLabelFor(key: MediaMetadataKey.locationLongitude)
    case kGCKMetadataKeyLocationName:
        return getEnumLabelFor(key: MediaMetadataKey.locationName)
    case kGCKMetadataKeyQueueItemID:
        return getEnumLabelFor(key: MediaMetadataKey.queueItemId)
    case kGCKMetadataKeyReleaseDate:
        return getEnumLabelFor(key: MediaMetadataKey.releaseDate)
    case kGCKMetadataKeySeasonNumber:
        return getEnumLabelFor(key: MediaMetadataKey.seasonNumber)
    case kGCKMetadataKeySectionDuration:
        return getEnumLabelFor(key: MediaMetadataKey.sectionDuration)
    case kGCKMetadataKeySectionStartAbsoluteTime:
        return getEnumLabelFor(key: MediaMetadataKey.sectionStartAbsoluteTime)
    case kGCKMetadataKeySectionStartTimeInContainer:
        return getEnumLabelFor(key: MediaMetadataKey.sectionStartTimeInContainer)
    case kGCKMetadataKeySectionStartTimeInMedia:
        return getEnumLabelFor(key: MediaMetadataKey.sectionStartTimeInMedia)
    case kGCKMetadataKeySeriesTitle:
        return getEnumLabelFor(key: MediaMetadataKey.seriesTitle)
    case kGCKMetadataKeyStudio:
        return getEnumLabelFor(key: MediaMetadataKey.studio)
    case kGCKMetadataKeySubtitle:
        return getEnumLabelFor(key: MediaMetadataKey.subtitle)
    case kGCKMetadataKeyTitle:
        return getEnumLabelFor(key: MediaMetadataKey.title)
    case kGCKMetadataKeyTrackNumber:
        return getEnumLabelFor(key: MediaMetadataKey.trackNumber)
    case kGCKMetadataKeyWidth:
        return getEnumLabelFor(key: MediaMetadataKey.width)
    default:
        return "default"
    }
}

func getFlutterMediaType(mediaType: GCKMediaMetadataType?) -> MediaType {
    switch mediaType {
    case .generic:
        return MediaType.generic
    case .movie:
        return MediaType.movie
    case .tvShow:
        return MediaType.tvShow
    case .musicTrack:
        return MediaType.musicTrack
    case .photo:
        return MediaType.photo
    case .audioBookChapter:
        return MediaType.audiobookChapter
    case .user:
        return MediaType.user
    case .none:
        return MediaType.generic
    @unknown default:
        return MediaType.generic
    }
}

func getFlutterMediaQueueItem(item: GCKMediaQueueItem?) -> MediaQueueItem {
    let result = MediaQueueItem()
    
    if (item != nil) {
        result.itemId = NSNumber(value: item?.itemID ?? 0)
    }
    
    result.autoplay = NSNumber(value: item?.autoplay ?? false)
    result.playbackDuration = NSNumber(value: item?.playbackDuration ?? -1)
    result.startTime = NSNumber(value: item?.startTime ?? 0)
    result.preloadTime = NSNumber(value: item?.preloadTime ?? 0)
    result.media = getFlutterMediaInfo(mediaInfo: item?.mediaInformation)
    
    return result
}

File: android/src/main/kotlin/com/gianlucaparadise/flutter_cast_framework/media/FlutterMediaLoadRequestDataHelper.kt
package com.gianlucaparadise.flutter_cast_framework.media

import com.gianlucaparadise.flutter_cast_framework.PlatformBridgeApis
import com.google.android.gms.cast.*
import com.google.android.gms.common.images.WebImage

fun getFlutterMediaStatus(mediaStatus: MediaStatus?): PlatformBridgeApis.MediaStatus {
    val flutterMediaInfo = getFlutterMediaInfo(mediaStatus?.mediaInfo)
    val flutterPlayerState = getFlutterPlayerState(mediaStatus?.playerState)
    val flutterAdBreakStatus = getFlutterAdBreakStatus(mediaStatus?.adBreakStatus)

    return PlatformBridgeApis.MediaStatus().apply {
        isPlayingAd = mediaStatus?.isPlayingAd ?: false
        mediaInfo = flutterMediaInfo
        playerState = flutterPlayerState
        adBreakStatus = flutterAdBreakStatus
    }
}

fun getFlutterAdBreakStatus(adBreakStatus: AdBreakStatus?): PlatformBridgeApis.AdBreakStatus {
    return PlatformBridgeApis.AdBreakStatus().apply {
        adBreakId = adBreakStatus?.breakId ?: ""
        adBreakClipId = adBreakStatus?.breakClipId ?: ""
        whenSkippableMs = adBreakStatus?.whenSkippableInMs ?: -1
    }
}

fun getFlutterPlayerState(playerStateRaw: Int?): PlatformBridgeApis.PlayerState {
    return when (playerStateRaw) {
        MediaStatus.PLAYER_STATE_UNKNOWN -> PlatformBridgeApis.PlayerState.UNKNOWN
        MediaStatus.PLAYER_STATE_BUFFERING -> PlatformBridgeApis.PlayerState.BUFFERING
        MediaStatus.PLAYER_STATE_IDLE -> PlatformBridgeApis.PlayerState.IDLE
        MediaStatus.PLAYER_STATE_LOADING -> PlatformBridgeApis.PlayerState.LOADING
        MediaStatus.PLAYER_STATE_PAUSED -> PlatformBridgeApis.PlayerState.PAUSED
        MediaStatus.PLAYER_STATE_PLAYING -> PlatformBridgeApis.PlayerState.PLAYING
        else -> PlatformBridgeApis.PlayerState.UNKNOWN
    }
}

fun getFlutterMediaInfo(mediaInfo: MediaInfo?): PlatformBridgeApis.MediaInfo {
    val flutterMediaMetadata = getFlutterMediaMetadata(mediaInfo?.metadata)
    val flutterMediaTracks = getFlutterMediaTracks(mediaInfo?.mediaTracks)
    val flutterStreamType = getFlutterStreamType(mediaInfo?.streamType)
    val flutterAdBreakClips = getFlutterAdBreakClips(mediaInfo?.adBreakClips)

    return PlatformBridgeApis.MediaInfo().apply {
        contentId = mediaInfo?.contentId ?: ""
        contentType = mediaInfo?.contentType ?: ""
        customDataAsJson = mediaInfo?.customData?.toString()
        mediaMetadata = flutterMediaMetadata
        mediaTracks = flutterMediaTracks
        streamDuration = mediaInfo?.streamDuration ?: 0
        streamType = flutterStreamType
        adBreakClips = flutterAdBreakClips
    }
}

fun getFlutterStreamType(streamType: Int?): PlatformBridgeApis.StreamType {
    return when (streamType) {
        -1 -> PlatformBridgeApis.StreamType.INVALID
        0 -> PlatformBridgeApis.StreamType.NONE
        1 -> PlatformBridgeApis.StreamType.BUFFERED
        2 -> PlatformBridgeApis.StreamType.LIVE
        else -> PlatformBridgeApis.StreamType.INVALID
    }
}

fun getFlutterMediaTracks(mediaTracks: List<MediaTrack>?): List<PlatformBridgeApis.MediaTrack> {
    return mediaTracks?.map {
        getFlutterMediaTrack(it)
    } ?: emptyList()
}

fun getFlutterMediaTrack(mediaTrack: MediaTrack?): PlatformBridgeApis.MediaTrack {
    val flutterSubtype = getFlutterSubtype(mediaTrack?.subtype)
    val flutterType = getFlutterType(mediaTrack?.type)

    return PlatformBridgeApis.MediaTrack().apply {
        contentId = mediaTrack?.contentId ?: ""
        id = mediaTrack?.id ?: -1
        language = mediaTrack?.language ?: ""
        name = mediaTrack?.name ?: ""
        trackSubtype = flutterSubtype
        trackType = flutterType
    }
}

fun getFlutterAdBreakClips(adBreakClips: List<AdBreakClipInfo>?): List<PlatformBridgeApis.AdBreakClipInfo> {
    return adBreakClips?.map {
        getFlutterAdBreakClipInfo(it)
    } ?: emptyList()
}

fun getFlutterAdBreakClipInfo(adBreakClipInfo: AdBreakClipInfo?): PlatformBridgeApis.AdBreakClipInfo {
    return PlatformBridgeApis.AdBreakClipInfo().apply {
        id = adBreakClipInfo?.id
        title = adBreakClipInfo?.title
        contentId = adBreakClipInfo?.contentId
        contentUrl = adBreakClipInfo?.contentUrl
        clickThroughUrl = adBreakClipInfo?.clickThroughUrl
        durationMs = adBreakClipInfo?.durationInMs
        imageUrl = adBreakClipInfo?.imageUrl
        mimeType = adBreakClipInfo?.mimeType
        whenSkippableMs = adBreakClipInfo?.whenSkippableInMs
    }
}

fun getFlutterType(type: Int?): PlatformBridgeApis.TrackType {
    return when (type) {
        0 -> PlatformBridgeApis.TrackType.UNKNOWN
        1 -> PlatformBridgeApis.TrackType.TEXT
        2 -> PlatformBridgeApis.TrackType.AUDIO
        3 -> PlatformBridgeApis.TrackType.VIDEO
        else -> PlatformBridgeApis.TrackType.UNKNOWN
    }
}

fun getFlutterSubtype(subtype: Int?): PlatformBridgeApis.TrackSubtype {
    return when (subtype) {
        -1 -> PlatformBridgeApis.TrackSubtype.UNKNOWN
        0 -> PlatformBridgeApis.TrackSubtype.NONE
        1 -> PlatformBridgeApis.TrackSubtype.SUBTITLES
        2 -> PlatformBridgeApis.TrackSubtype.CAPTIONS
        3 -> PlatformBridgeApis.TrackSubtype.DESCRIPTIONS
        4 -> PlatformBridgeApis.TrackSubtype.CHAPTERS
        5 -> PlatformBridgeApis.TrackSubtype.METADATA
        else -> PlatformBridgeApis.TrackSubtype.UNKNOWN
    }
}

fun getFlutterMediaMetadata(mediaMetadata: MediaMetadata?): PlatformBridgeApis.MediaMetadata {
    val flutterMediaType = getFlutterMediaType(mediaMetadata?.mediaType)
    val flutterWebImages = getFlutterWebImages(mediaMetadata?.images)
    val flutterStrings = getFlutterStrings(mediaMetadata)

    return PlatformBridgeApis.MediaMetadata().apply {
        mediaType = flutterMediaType
        webImages = flutterWebImages
        strings = flutterStrings
    }
}

fun getFlutterWebImages(images: List<WebImage>?): List<PlatformBridgeApis.WebImage> {
    return images?.map {
        PlatformBridgeApis.WebImage().apply {
            url = it.url.toString()
        }
    } ?: emptyList()
}

fun getFlutterMediaType(mediaType: Int?): PlatformBridgeApis.MediaType {
    return when (mediaType) {
        0 -> PlatformBridgeApis.MediaType.GENERIC
        1 -> PlatformBridgeApis.MediaType.MOVIE
        2 -> PlatformBridgeApis.MediaType.TV_SHOW
        3 -> PlatformBridgeApis.MediaType.MUSIC_TRACK
        4 -> PlatformBridgeApis.MediaType.PHOTO
        5 -> PlatformBridgeApis.MediaType.AUDIOBOOK_CHAPTER
        100 -> PlatformBridgeApis.MediaType.USER
        else -> PlatformBridgeApis.MediaType.GENERIC
    }
}

fun getFlutterStrings(mediaMetadata: MediaMetadata?): Map<String, String?> {
    val stringsKeys = mediaMetadata?.keySet() ?: return emptyMap()
    return stringsKeys.associate { getFlutterMediaMetadataKey(it) to mediaMetadata.getString(it) }
}

fun getFlutterMediaMetadataKey(mediaMetadataKey: String): String {
    return when (mediaMetadataKey) {
        "com.google.android.gms.cast.metadata.ALBUM_ARTIST" -> PlatformBridgeApis.MediaMetadataKey.ALBUM_ARTIST.name
        "com.google.android.gms.cast.metadata.ALBUM_TITLE" -> PlatformBridgeApis.MediaMetadataKey.ALBUM_TITLE.name
        "com.google.android.gms.cast.metadata.ARTIST" -> PlatformBridgeApis.MediaMetadataKey.ARTIST.name
        "com.google.android.gms.cast.metadata.BOOK_TITLE" -> PlatformBridgeApis.MediaMetadataKey.BOOK_TITLE.name
        "com.google.android.gms.cast.metadata.BROADCAST_DATE" -> PlatformBridgeApis.MediaMetadataKey.BROADCAST_DATE.name
        "com.google.android.gms.cast.metadata.CHAPTER_NUMBER" -> PlatformBridgeApis.MediaMetadataKey.CHAPTER_NUMBER.name
        "com.google.android.gms.cast.metadata.CHAPTER_TITLE" -> PlatformBridgeApis.MediaMetadataKey.CHAPTER_TITLE.name
        "com.google.android.gms.cast.metadata.COMPOSER" -> PlatformBridgeApis.MediaMetadataKey.COMPOSER.name
        "com.google.android.gms.cast.metadata.CREATION_DATE" -> PlatformBridgeApis.MediaMetadataKey.CREATION_DATE.name
        "com.google.android.gms.cast.metadata.DISC_NUMBER" -> PlatformBridgeApis.MediaMetadataKey.DISC_NUMBER.name
        "com.google.android.gms.cast.metadata.EPISODE_NUMBER" -> PlatformBridgeApis.MediaMetadataKey.EPISODE_NUMBER.name
        "com.google.android.gms.cast.metadata.HEIGHT" -> PlatformBridgeApis.MediaMetadataKey.HEIGHT.name
        "com.google.android.gms.cast.metadata.LOCATION_LATITUDE" -> PlatformBridgeApis.MediaMetadataKey.LOCATION_LATITUDE.name
        "com.google.android.gms.cast.metadata.LOCATION_LONGITUDE" -> PlatformBridgeApis.MediaMetadataKey.LOCATION_LONGITUDE.name
        "com.google.android.gms.cast.metadata.LOCATION_NAME" -> PlatformBridgeApis.MediaMetadataKey.LOCATION_NAME.name
        "com.google.android.gms.cast.metadata.QUEUE_ITEM_ID" -> PlatformBridgeApis.MediaMetadataKey.QUEUE_ITEM_ID.name
        "com.google.android.gms.cast.metadata.RELEASE_DATE" -> PlatformBridgeApis.MediaMetadataKey.RELEASE_DATE.name
        "com.google.android.gms.cast.metadata.SEASON_NUMBER" -> PlatformBridgeApis.MediaMetadataKey.SEASON_NUMBER.name
        "com.google.android.gms.cast.metadata.SECTION_DURATION" -> PlatformBridgeApis.MediaMetadataKey.SECTION_DURATION.name
        "com.google.android.gms.cast.metadata.SECTION_START_ABSOLUTE_TIME" -> PlatformBridgeApis.MediaMetadataKey.SECTION_START_ABSOLUTE_TIME.name
        "com.google.android.gms.cast.metadata.SECTION_START_TIME_IN_CONTAINER" -> PlatformBridgeApis.MediaMetadataKey.SECTION_START_TIME_IN_CONTAINER.name
        "com.google.android.gms.cast.metadata.SECTION_START_TIME_IN_MEDIA" -> PlatformBridgeApis.MediaMetadataKey.SECTION_START_TIME_IN_MEDIA.name
        "com.google.android.gms.cast.metadata.SERIES_TITLE" -> PlatformBridgeApis.MediaMetadataKey.SERIES_TITLE.name
        "com.google.android.gms.cast.metadata.STUDIO" -> PlatformBridgeApis.MediaMetadataKey.STUDIO.name
        "com.google.android.gms.cast.metadata.SUBTITLE" -> PlatformBridgeApis.MediaMetadataKey.SUBTITLE.name
        "com.google.android.gms.cast.metadata.TITLE" -> PlatformBridgeApis.MediaMetadataKey.TITLE.name
        "com.google.android.gms.cast.metadata.TRACK_NUMBER" -> PlatformBridgeApis.MediaMetadataKey.TRACK_NUMBER.name
        "com.google.android.gms.cast.metadata.WIDTH" -> PlatformBridgeApis.MediaMetadataKey.WIDTH.name
        else -> mediaMetadataKey
    }
}

fun getFlutterMediaQueueItem(item: MediaQueueItem?): PlatformBridgeApis.MediaQueueItem {
    val mediaInto = getFlutterMediaInfo(item?.media)

    return PlatformBridgeApis.MediaQueueItem().apply {
        itemId = item?.itemId?.toLong()
        autoplay = item?.autoplay ?: false
        playbackDuration = item?.playbackDuration ?: -1.0
        startTime = item?.startTime ?: 0.0
        preloadTime = item?.preloadTime ?: 0.0
        media = mediaInto
    }
}